### PR TITLE
Cleanup Announce protocol: Own module + modularization of behaviour

### PR DESCRIPTION
**Pull request by hbandura**
This PR continues work to move out the AnnounceManager from the istanbul.Backend type.

The most important part is removing any Backend dependencies from the AnnounceManager, thus paving the way for better testing of all the announce protocol functionality.

The background announce thread is modeled in the AnnounceWorker type. While there is an AnnounceTaskState, it's not really a state machine state, so the naming is not ideal and it's open for suggestions.

Many small objects were created that have a mapping 1:1 to the previous functions in the announce code: therefore some don't have a well thought responsibility but exist as they are to ease the transition from the previous/existing code.

Moving all of it to its own package proved to be a bit hard and would make this PR much more confusing, so I plan to do that in a different PR since it would require tinkering with some types like the protocol messages. All of the files that I plan to move out are prepended by the name `announce_`.

### CODE REVIEW

**Comment by mcortesi** [comment]
need to still read announce and announc_task

File: consensus/istanbul/backend/backend.go [hunk 1]
```diff
@@ -197,6 +194,12 @@ func New(config *istanbul.Config, db ethdb.Database) consensus.Istanbul {
 		IsProxiedValidator: backend.IsProxiedValidator(),
 		AWallets:           &backend.aWallets,
 		VcDbPath:           config.VersionCertificateDBPath,
+		Announce:           backend.config.Announce,
+		Epoch:              backend.config.Epoch,
+	}
+
+	peerCounter := func(purpose p2p.PurposeFlag) int {
```
[the remark this answers is not in the document]
Comment (hbandura) [reply consensus/istanbul/backend/backend.go:201]: You are right

File: consensus/istanbul/backend/announce_task.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import (
+	"time"
+
+	"github.com/celo-org/celo-blockchain/consensus/istanbul"
+	"github.com/celo-org/celo-blockchain/p2p"
+)
+
+// announceTaskState encapsulates the state needed to guide the behavior of the announce protocol
+// thread. This type is designed to be used from A SINGLE THREAD only.
+type announceTaskState struct {
```
Comment (mcortesi) [inline consensus/istanbul/backend/announce_task.go:12]: I like the ideas of modeling the protocol as a state machine. where you get "events" and react to them. Some events are "time", others are network messages, etc.

But seeing it like that, seems to make it simpler to understand

Comment (mcortesi) [reply consensus/istanbul/backend/announce_task.go:12]: still... not sure the separation of announceManager and announceTask state. My sense is that it explores this "state machine" modeling, but the announce manager is also the protocol...

File: consensus/istanbul/config.go [hunk 3]
```diff
@@ -80,26 +83,28 @@ type ProxyConfig struct {
...
+	DefaultLookbackWindow:       12,
+	ReplicaStateDBPath:          "replicastate",
+	ValidatorEnodeDBPath:        "validatorenodes",
+	VersionCertificateDBPath:    "versioncertificates",
+	RoundStateDBPath:            "roundstates",
+	Validator:                   false,
+	Replica:                     false,
+	Proxy:                       false,
+	Proxied:                     false,
+	Announce: &AnnounceConfig{
```
[the remark this answers is not in the document]
Comment (hbandura) [reply consensus/istanbul/config.go:102]: Oh, I completely missed that.

If that were the case, what do you think should I do moving forward? I do think that having Announce parameters in a type by itself is an improvement if possible

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 0db4c9ed8d70b419816a64b7df3215b4d4a76588
<details><summary>coverage:  44.5% of statements across all listed packages</summary><br>
<pre>
coverage:  54.1% of statements in consensus/istanbul
coverage:  45.3% of statements in consensus/istanbul/backend
coverage:  10.5% of statements in consensus/istanbul/backend/announce
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.4% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: 9989c80757
</details>

**Comment by piersy** [comment]
Hi @hbandura,

I've had a look through this and I can see where it's headed, but I don't think it would be helpful to merge it in its current state.

My reasoning for this is as follows:
* This PR adds a significant amount of code, an extra 750 lines after accounting for comments. This adds to the burden of anyone wanting to understand this code, it's a lot of extra lines to read and understand.

* Although the new code separates the announce code from the backend package at a compile level at runtime the announce functionality is still totally dependent on the backend.

* The compile time separation of the announce code from the backend code involves the introduction of many new interfaces that add extra layers of indirection between the announce code and the backend, and all but one or two lead back to the backend. For example `AnnounceManager -> Worker -> AnnounceStatePruner -> backend.Backend`. The greater the number of layers of indirection the more hidden the root dependency becomes. 

* Hidden dependencies make it hard to refactor because they impart the false expectation that you can separate some part of the code, but after tracking down all the bits you realise that you can't perform the extraction you wanted because some element depends on something that isn't in scope at the point you want to construct your object. It essentially delays the realisation about what is extractable and that costs time to developers.

I'm actually facing this problem right now trying to provide an injectable message sender to the backend so that I can mock that out for testing. I separated the peer set from the p2p.Server, so the backend didn't need the broadcaster set on it after construction and so the message sender could just be constructed with a peer set. But then I needed access to the ValidatorEnodeDB to be able to look up validator peer ids, no problem just move that package out of internal, but then I needed the validatorPeerHandler to be able to construct the ValEnodeDB and validatorPeerHandler has a reference to the backend, so now I'm stuck, having spent some hours going down this path. Thats actually just one branch out of 5 that I currently have for this task where a similar thing has happened trying to follow a different path. Essentially all paths lead to the backend.

So while I think this PR represents a step in the right direction I see it as an intermediate step, and for the changes here to really make this area of code nicer to work with we need to break down the backend object so that the interfaces you have defined can be constructed without encountering circular dependencies, such as:

`backend.Backend -> AnnounceManager -> Worker -> AnnounceStatePruner -> backend.Backend`

I'd be very happy to work with you on this and I think we could come up with something that significantly reduces the amount of code added and also makes the istanbul package much nicer to work in.

I already have the separation of the Peers object, which I think could go a long way towards removing some of the circular dependencies existing here.
See 98fb51618810375f22214ce7a8fc4f15d0d3b046

Let me know what you think?

File: consensus/istanbul/announce/network.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package announce
+
+import "github.com/celo-org/celo-blockchain/common"
+
+type Network interface {
+	// Gossip gossips protocol messages
+	Gossip(payload []byte, ethMsgCode uint64) error
+	// RetrieveValidatorConnSet returns the validator connection set
+	RetrieveValidatorConnSet() (map[common.Address]bool, error)
```
[the remark this answers is not in the document]
Comment (mcortesi) [reply consensus/istanbul/announce/network.go:9]: this seems to be "NearlyElectedValidatorSet()", not connections

File: consensus/istanbul/announce/manager.go [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+			return err
+		}
+	}
+
+	// Upsert regardless to account for the case that the target is a non-ValidatorPurpose
+	// peer but should be.
+	// If the target is not a peer and should be a ValidatorPurpose peer, this
+	// will designate the target as a ValidatorPurpose peer and send an enodeCertificate
+	// during the istanbul handshake.
+	if err := m.state.ValEnodeTable.UpsertVersionAndEnode([]*istanbul.AddressEntry{{Address: address, Node: node, Version: version}}); err != nil {
```
Comment (mcortesi) [inline consensus/istanbul/announce/manager.go:264]: this seems out of place, this is not "answering" the msg, but processing it

**Review by mcortesi — COMMENTED** [review]
Left some comments while i continue with the review

**Review by mcortesi — COMMENTED** [review]
From 3 weeks ago. i'll finish the rest now

**Review by mcortesi — APPROVED** [review]
LGTM. On previous reviews there are many comments, none is blocker, can be done later.

**Comment by hbandura** [comment]
I will make the requested modifications in a following PR

**Pull request merged**